enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -608,51 +608,90 @@ class AdjointGenerator
+        Value *dif = diffe(&I, Builder2);
 
-    if (!gutils->isConstantValue(orig_op0)) {
-      Value *dif = diffe(&I, Builder2);
-
-      size_t size = 1;
-      if (orig_op0->getType()->isSized())
-        size = (gutils->newFunc->getParent()->getDataLayout().getTypeSizeInBits(
-                    orig_op0->getType()) +
-                7) /
-               8;
-      Type *FT = TR.addingType(size, orig_op0);
-      if (!FT) {
-        llvm::errs() << " " << *gutils->oldFunc << "\n";
-        TR.dump();
-        llvm::errs() << " " << *orig_op0 << "\n";
+        size_t size = 1;
+        if (orig_op0->getType()->isSized())
+          size =
+              (gutils->newFunc->getParent()->getDataLayout().getTypeSizeInBits(
+                   orig_op0->getType()) +
+               7) /
+              8;
+        Type *FT = TR.addingType(size, orig_op0);
+        if (!FT) {
+          llvm::errs() << " " << *gutils->oldFunc << "\n";
+          TR.dump();
+          llvm::errs() << " " << *orig_op0 << "\n";
+        }
+        assert(FT);
+        if (I.getOpcode() == CastInst::CastOps::FPTrunc ||
+            I.getOpcode() == CastInst::CastOps::FPExt) {
+          addToDiffe(orig_op0, Builder2.CreateFPCast(dif, op0->getType()),
+                     Builder2, FT);
+        } else if (I.getOpcode() == CastInst::CastOps::BitCast) {
+          addToDiffe(orig_op0, Builder2.CreateBitCast(dif, op0->getType()),
+                     Builder2, FT);
+        } else if (I.getOpcode() == CastInst::CastOps::Trunc) {
+          // TODO CHECK THIS
+          auto trunced = Builder2.CreateZExt(dif, op0->getType());
+          addToDiffe(orig_op0, trunced, Builder2, FT);
+        } else {
+          TR.dump();
+          llvm::errs() << *I.getParent()->getParent() << "\n"
+                       << *I.getParent() << "\n";
+          llvm::errs() << "cannot handle above cast " << I << "\n";
+          report_fatal_error("unknown instruction");
+        }
       }
-      assert(FT);
-      if (I.getOpcode() == CastInst::CastOps::FPTrunc ||
-          I.getOpcode() == CastInst::CastOps::FPExt) {
-        addToDiffe(orig_op0, Builder2.CreateFPCast(dif, op0->getType()),
-                   Builder2, FT);
-      } else if (I.getOpcode() == CastInst::CastOps::BitCast) {
-        addToDiffe(orig_op0, Builder2.CreateBitCast(dif, op0->getType()),
-                   Builder2, FT);
-      } else if (I.getOpcode() == CastInst::CastOps::Trunc) {
-        // TODO CHECK THIS
-        auto trunced = Builder2.CreateZExt(dif, op0->getType());
-        addToDiffe(orig_op0, trunced, Builder2, FT);
+      setDiffe(&I, Constant::getNullValue(I.getType()), Builder2);
+
+      break;
+    }
+    case DerivativeMode::ForwardMode: {
+      Value *orig_op0 = I.getOperand(0);
+
+      IRBuilder<> Builder2(&I);
+      getForwardBuilder(Builder2);
+
+      if (!gutils->isConstantValue(orig_op0)) {
+        Value *dif = diffe(orig_op0, Builder2);
```

Comment:
This still should be able to be simplified. E.g. it could have a single case like setDiffe(&I, Builder.CreateCast(diff, I.getType(), I.getOpcode())

Reply:
done